Controlled Qutrit Unitary

mudit2812 opened the pull request:
**Context:**
There are very few simulators available today that offer qudit simulation. One notable example is [Google's Cirq](https://quantumai.google/cirq/qudits), which includes support for qudit operations. However, the simulator is far from complete, and a key drawback is the lack of operations available, which creates a need for users to define operations themselves. Popular libraries such as Qiskit and QuTiP also include some qudit manipulation capabilities, but with few use cases. Beyond these libraries, qudit simulation solutions are quite limited. Moreover, qudit simulation is a feature that is already in demand, as seen in this [issue](https://github.com/PennyLaneAI/pennylane/issues/2190) on the PennyLane repository.

This PR builds upon the previous PR's that added qutrit functionality by implementing the controlled `QutritUnitary` operation

**Description of the Change:**
* Added `ControlledQutritUnitary()` operation to PennyLane to use with qutrit devices.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/2190

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -7,6 +7,10 @@
+  
+* Added the `qml.ControlledQutritUnitary` qutrit operation for applying a controlled arbitrary unitary matrix to the specified set of wires.
+Users can specify the control wires as well as the values to control the operation on.
+  ([#2844](https://github.com/PennyLaneAI/pennylane/pull/2844))
```
This might be a good place to include a usage example, since the control values are now ternary, and the default control is 2 instead of 1.

glassnotes commented on doc/releases/changelog-dev.md:
```diff
@@ -7,6 +7,10 @@
 keyword argument when using `GellMann`, which determines which of the 8 Gell-Mann matrices is used as the observable.
   ([#3035](https://github.com/PennyLaneAI/pennylane/pull/3035))
+  
+* Added the `qml.ControlledQutritUnitary` qutrit operation for applying a controlled arbitrary unitary matrix to the specified set of wires.
```
Please add the new operation to the list in `docs/introduction/operations.rst`.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,214 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+    wire ``3`` conditioned on three wires where the first is in state ``0``, the
+    second is in state ``1``, and the third in state ``2``, we can write:
+
+    >>> qml.ControlledQutritUnitary(U, control_wires=[0, 1, 2], wires=3, control_values='012')
```
Great example :+1: Rendering in the docs looks good.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,214 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+            )
+
+        self._hyperparameters = {
+            "u_wires": wires,
```
What is `u_wires`?

mudit2812 replied:
`u_wires` is what the name of the hyperparameter was in `ControlledQubitUnitary` for the target wires, so I used the same name for the hyperparameter in `ControlledQutritUnitary`.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,214 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+        Args:
+            U (tensor_like): unitary matrix
+            control_wires (Iterable): the control wire(s)
+            u_wires (Iterable): the wire(s) the unitary acts on
```
Oh I see, has the argument been renamed for the controlled operations from `wires` to `u_wires`?

mudit2812 replied:
For this function, the parameter is called `u_wires` because the parameters for `compute_matrix` use the same names as the operation parameters/hyperparameters.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,214 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+                )
+
+            # Make sure all values are either 0 or 1 or 2
+            if any(x not in ["0", "1", "2"] for x in control_values):
```
Could use a set here as is done for `ControlledQubitUnitary`, it helps the logic look a bit clearer.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,214 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+            ]
+        return super().pow(z)
+
+    def _controlled(self, wire):
```
I notice this looks a bit different than the same method in `ControlledQubitUnitary`, we're not considering any inverses here. Is there a reason this is different?

mudit2812 replied:
`_controlled` for `ControlledQubitUnitary` used to be different. I think it got updated recently because all the design decisions I made for `ControlledQutritUnitary` were based on `ControlledQubitUnitary`.

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
         assert np.allclose(res_dynamic, expected)
 
+    def test_matrix_representation_broadcasted(self):
+        """Test that the matrix representation is defined correctly"""
```
This docstring matches the previous one (please update). Also the test looks almost the same except for the `tensordot` line; is that what's applying the broadcasting?

mudit2812 replied:
Yes, that's how the broadcasting is being added to the test.

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        assert np.allclose(res_static, expected)
+        assert np.allclose(res_dynamic, expected)
+
+    def test_controlled(self):
```
I recommend splitting up this test into a couple smaller ones that address a few different ways of applying a controlled `QutritUnitary`:
 - the way you've already done so here with a single-qutrit gate
 - controlling a two-qutrit `QutritUnitary` operation 
 - using `qml.ctrl` instead of `op._controlled`
 - controlling in a case where the control qutrit is in a superposition
 - a different control value than 2

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        state_1 = f1()
+        state_2 = f2()
+
+        assert np.allclose(state_1, state_2)
```
I feel like this test needs a test :sweat_smile: I was able to follow though, and this is a good sanity check to make sure things work under wire reordering.

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        U = unitary_group.rvs(3 ** len(target_wires), random_state=1967)
+
+        # Pick random starting state for the control and target qutrits
+        starting_state = np.random.normal(size=3 ** (len(control_wires + target_wires)))
```
I think it wouldn't hurt here to make the state contain complex values just to cover our bases.

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        # Pick random starting state for the control and target qutrits
+        starting_state = np.random.normal(size=3 ** (len(control_wires + target_wires)))
+        starting_state = starting_state / np.linalg.norm(starting_state)
+        print(np.linalg.norm(starting_state))
```
```suggestion

```

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        one_locations = [x for x in range(len(control_values)) if control_values[x] == "1"]
+        zero_locations = [x for x in range(len(control_values)) if control_values[x] == "0"]
+
+        ops = []
```
I think I followed what's happening here (you are manually applying shifts to set control values, apply the operation, and then undoing them)? It would be helpful to add some developer-facing comments to explain the process a little more.

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        assert np.allclose(res_dynamic, expected, atol=tol)
+
+    def test_matrix_representation_broadcasted(self, tol):
+        """Test that the matrix representation is defined correctly"""
```
```suggestion
        """Test that the matrix representation is defined correctly when broadcasting"""
```

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        assert np.allclose(res_dynamic, expected, atol=tol)
+
+    def test_no_decomp(self):
+        """Test that ControlledQutritUnitary raises a decomposition undefined
+        error."""
```
```suggestion
        """Test that ControlledQutritUnitary raises a DecompositionUndefinedError."""
```

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+            op.pow(0.12)
+
+    def test_noninteger_pow_broadcasted(self):
+        """Test that a ControlledQutritUnitary raised to a non-integer power raises an error."""
```
```suggestion
        """Test that a ControlledQutritUnitary raised to a non-integer power raises an error when broadcasting."""
```

glassnotes commented on tests/ops/qutrit/test_qutrit_matrix_ops.py:
```diff
@@ -283,9 +283,363 @@ def test_matrix_representation(self):
+        with pytest.raises(qml.operation.PowUndefinedError):
+            op.pow(0.12)
+
+    def test_controlled(self):
```
Please add a docstring and make the test name a bit more descriptive; is this testing a controlled `ControlledQutritUnitary`? If so I recommend to do something similar as the `QutritUnitary` case, where doing a controlled operation is tested in a few different ways, like with `qml.ctrl`.

mudit2812 replied:
I added a few test cases that use `qml.ctrl` for `TestQutritUnitary.test_controlled`, but I'm keeping them commented out for now and am a bit hesitant to add tests for `qml.ctrl` for testing `ControlledQutritUnitary` because `qml.ctrl` is yet another function hard-coded to work with qubits, as any control values higher than 1 get treated as booleans so that they can be changed to 0 or 1, therefore not allowing the use of 2 as a control value.

mudit2812 replied:
I've added a feature request. [Here it is](https://github.com/PennyLaneAI/pennylane/issues/3114).

glassnotes replied:
Thanks @mudit2812 !

glassnotes reviewed (changes requested):
@mudit2812 looking good! Mostly just some suggestions for restructuring a few of the tests here.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -316,7 +317,8 @@ def pow(self, z):
-        ctrl_wires = sorted(self.control_wires + wire)
-        ControlledQutritUnitary(
+        ctrl_wires = self.control_wires + wire
+        new_op = ControlledQutritUnitary(
```
Just taking a look at the [qubit case](https://github.com/PennyLaneAI/pennylane/blob/d565eec3c22d45aff1bec1029134aafe0d4f50bb/pennylane/ops/qubit/matrix_ops.py#L378), I think we will need to update the `control_values` as well here to ensure that they are preserved when the new operation is constructed. This change will also need a corresponding test.

glassnotes commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,216 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+    available for ``ControlledQutritUnitary``:
+
+    * ``control_wires``: wires that act as control for the operation
+    * ``U``: unitary applied to the target wires
```
```suggestion
    * ``U``: unitary applied to the target wires
    * ``control_values``: a string of trits representing the state of the control
                qutrits to control on (default is the all 2s state)    
```

glassnotes reviewed (changes requested):
@mudit2812 changes look good. Really only one suggestion, to take into account mixed polarity controls upon controlled `ControlledQutritUnitary` construction.

albi3ro commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,223 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+        super().__init__(*params, wires=total_wires, do_queue=do_queue)
+
+    @staticmethod
+    def compute_decomposition(*params, wires=None, **hyperparameters):
```
Should just inherit this behaviour from the base class.  We don't need to redefine it here.

mudit2812 commented:
Hi @albi3ro ! Thanks for the review. I do agree about the note you made regarding the significant overlap and redundancy between `QubitUnitary`, `QutritUnitary`, and `ControlledQutritUnitary`. When this PR was originally created, `ControlledQutritUnitary` was the best option due to the incomplete `qml.op_math` module. However, now that that isn't the case, `qml.ctrl` (and other functions) can be updated to support qutrit operations, which is my next goal.

With that being said, I've addressed the changes you suggested in your review. I've added `ControlledQutritUnitary.adjoint()`. I also noticed that neither `ControlledQubitUnitary` nor `ControlledQutritUnitary` store the control values in the `pow` method, thus the operation returned by that method doesn't include the correct control values. I've updated `ControlledQutritUnitary.pow()` to work correctly.

glassnotes commented:
> I also noticed that neither `ControlledQubitUnitary` nor `ControlledQutritUnitary` store the control values in the `pow` method, thus the operation returned by that method doesn't include the correct control values. I've updated `ControlledQutritUnitary.pow()` to work correctly.

@mudit2812 that sounds like a bug :fearful:  could you please log an issue for that (if there isn't one already)?

glassnotes reviewed (approved):
Thanks @mudit2812, updated version looks good to me. 

I agree with @albi3ro regarding the inheritance structure, I didn't notice this before. I'm inclined, for consistency (but at the cost of additional debt), to leave as is since it matches `QubitUnitary`/`ControlledQubitUnitary`, and then the structure of both can be updated together in a consistent manner - does that sound reasonable?

albi3ro commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,232 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+    :math:`\vert 0\rangle` or :math:`\vert 1\rangle` state, or a mix of the three.
+
+    The state on which to control can be changed by passing a string of trits to
+    `control_values`. For example, if we want to apply a single-qutrit unitary to
```
```suggestion
    ``control_values``. For example, if we want to apply a single-qutrit unitary to
```

josh146 commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,232 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+
+    * ``control_wires``: wires that act as control for the operation
+    * ``U``: unitary applied to the target wires
+    * ``control_values``: a string of trits representing the state of the control
+                qutrits to control on (default is the all 2s state)
```
```suggestion
    * ``control_values``: a string of trits representing the state of the control
      qutrits to control on (default is the all 2s state)
```

albi3ro commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,232 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+    r"""ControlledQutritUnitary(U, control_wires, wires, control_values)
+    Apply an arbitrary fixed unitary to ``wires`` with control from the ``control_wires``.
+
+    In addition to default ``Operation`` instance attributes, the following are
+    available for ``ControlledQutritUnitary``:
```
This sentence makes it sound like `U` and `control_values` are "attributes", and therefore accessible like `op.U` and `op.control_values`. Should we make them properties? 

Or we can just remove this blurb. The available attributes are accessible under "Attributes" in the documentation anyway.

mudit2812 replied:
I see your point. I do see the benefit of highlighting which attributes/properties are available to users, but as you said, it is a bit redundant. I think that making `control_values` a property is helpful definitely for the sake of convenient access, but I'm not entirely sure if accessing `op.U` makes much sense as in most use cases users will likely already have access to the unitary. I've added `control_values` as a property for now. Let me know what you think!

albi3ro replied:
Thanks for adding the property.  I added a suggestion to clarify that `op.U` itself isn't an attribute, but accessible via `op.parameters[0]`.

albi3ro commented on pennylane/ops/qutrit/matrix_ops.py:
```diff
@@ -98,20 +109,234 @@ def compute_matrix(U):  # pylint: disable=arguments-differ
+    available for ``ControlledQutritUnitary``:
+
+    * ``control_wires``: wires that act as control for the operation
+    * ``U``: unitary applied to the target wires
```
```suggestion
    * ``U``: unitary applied to the target wires. Accessible via ``op.parameters[0]``
```

albi3ro reviewed (approved):
Thanks for addressing the comments 👍
